src/Evolution/Systems/Cce/NewmanPenrose.hpp

```diff
@@ -63,4 +71,123 @@ struct VolumeWeyl<Tags::Psi0> {
+          cauchy_view_volume_j,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& gauge_cauchy_c,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& volume_j,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& gauge_cauchy_d,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& omega_cauchy,
+      const Spectral::Swsh::SwshInterpolator& interpolator,
+      const size_t l_max);
+};
+
+/*!
+ * \brief Compute the Weyl scalar \f$\Psi_0\f$ in the volume for the purpose
+ * of CCM, the quantity is in the Cauchy coordinates.
+ *
+ * \details The Weyl scalar \f$\Psi_0\f$ is given by:
+ *
+ * \f{align*}{
+ * \Psi_0 = \frac{(1 - y)^4}{16 r^2 K}
+ * \bigg[& \partial_y \beta \left((1 + K) (\partial_y J)
+ * - \frac{J^2 \partial_y \bar J}{1 + K}\right)
+ * - \frac{1}{2} (1 + K) (\partial_y^2 J)
+ * + \frac{J^2 \partial_y^2 \bar J}{2(K + 1)}\\
+ * & + \frac{1}{K^2} \left(- \frac{1}{4} J \left(\bar{J}^2 \left(\partial_y
+ * J\right)^2 + J^2 \left(\partial_y \bar J\right)^2\right)
+ * + \frac{1 + K^2}{2} J (\partial_y J) (\partial_y \bar J)
+ * \right)\bigg].
+ * \f}
+ *
+ * The quantities above are all in the Cauchy coordinates, where \f$K\f$ is
+ * updated from \f$J\f$ and \f$\bar J\f$, \f$(1-y)\f$ is invariant under
+ * the coordinate transformation. \f$r\f$ transforms as
+ *
+ * \f{align*}{
+ * r = \omega \hat r
+ * \f}
+ */
+template <>
+struct VolumeWeyl<Tags::Psi0Match> {
+  using return_tags = tmpl::list<Tags::Psi0Match>;
+  using argument_tags =
+      tmpl::list<Tags::BondiJCauchyView, Tags::Dy<Tags::BondiJCauchyView>,
+                 Tags::Dy<Tags::Dy<Tags::BondiJCauchyView>>,
+                 Tags::BoundaryValue<Tags::BondiR>, Tags::OneMinusY,
+                 Tags::LMax>;
+  static void apply(
+      gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 2>>*> psi_0,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& bondi_j_cauchy,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& dy_j_cauchy,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& dy_dy_j_cauchy,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& bondi_r_cauchy,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& one_minus_y,
+      const size_t l_max);
+};
+
+/*!
+ * \brief Compute the Weyl scalar \f$\Psi_0\f$ and its radial derivative
+ * \f$\partial_\lambda \Psi_0\f$ on the inner boundary of CCE domain.
+ * The quantities are in the Cauchy coordinates.
+ *
+ * \details The radial derivative of the Weyl scalar \f$\partial_\lambda
+ * \Psi_0\f$ is given by
+ *
+ * \f{align*}{
+ * \partial_\lambda \Psi_0 = \frac{(1-y)^2}{2r}e^{-2\beta}
+ * \partial_y \Psi_0
+ * \f}
+ *
+ * Note that \f$(1-y)\f$, \f$r\f$, and \f$\beta\f$ are in the Cauchy
+ * coordinates, where \f$(1-y)\f$ is invariant under the coordinate
+ * transformation, while \f$r\f$ and \f$\beta\f$ transform as
+ *
+ * \f{align*}{
+ * &r = \omega \hat r
+ * & \beta = \hat \beta - \frac{1}{2} \log \omega
+ * \f}
+ */
+struct BoundaryWeyl {
```

Comment:
Do we want to call this `InnerBoundaryWeyl` since it operates only on the inner boundary?  (which is determined by the 0 that is passed into make_const_view in the cpp file).